tests/test/org/commcare/backend/session/test/SessionStackTests.java

```diff
@@ -85,21 +84,47 @@ public void testViewNav() throws Exception {
         CaseTestUtils.xpathEvalAndCompare(session.getEvaluationContext(),
                 "count(instance('session')/session/data/case_id_to_send)", "0");
 
-        Assert.assertEquals(session.getNeededData(), SessionFrame.STATE_COMMAND_ID);
+        Assert.assertEquals(SessionFrame.STATE_COMMAND_ID, session.getNeededData());
     }
 
     @Test
     public void testViewNonNav() throws Exception {
+        mApp = new MockApp("/complex_stack/");
         SessionWrapper session = mApp.getSession();
-        Assert.assertEquals(session.getNeededData(), SessionFrame.STATE_COMMAND_ID);
+
+        Assert.assertEquals(SessionFrame.STATE_COMMAND_ID, session.getNeededData());
 
         session.setCommand("m4-f0");
 
-        Assert.assertEquals(session.getNeededData(), SessionFrame.STATE_DATUM_VAL);
+        Assert.assertEquals(SessionFrame.STATE_DATUM_VAL, session.getNeededData());
 
-        Assert.assertEquals(session.getNeededDatum().getDataId(), "case_id_to_view");
+        Assert.assertEquals("case_id_to_view", session.getNeededDatum().getDataId());
 
         Assert.assertTrue("Session incorrectly tagged a view command", session.isViewCommand(session.getCommand()));
     }
 
+    @Test
+    public void testOutOfOrderStack() throws Exception {
+        mApp = new MockApp("/session-tests-template/");
+        SessionWrapper session = mApp.getSession();
+
+        // Select a form that has 3 datum requirements to enter (in order from suite.xml: case_id,
+        // case_id_new_visit_0, usercase_id)
+        Assert.assertEquals(SessionFrame.STATE_COMMAND_ID, session.getNeededData());
+        session.setCommand("m0");
+
+        Assert.assertEquals(SessionFrame.STATE_COMMAND_ID, session.getNeededData());
+        session.setCommand("m0-f3");
+
+        // Set 2 of the 3 needed datums, but not in order (1st and 3rd)
+        session.setDatum("case_id", "case_id_value");
+        session.setDatum("usercase_id", "usercase_id_value");
+
+        // Session should now need the case_id_new_visit_0, which is a computed datum
+        Assert.assertEquals(SessionFrame.STATE_DATUM_COMPUTED, session.getNeededData());
+
+        // The key of the needed datum should be "case_id_new_visit_0"
+        Assert.assertEquals("case_id_new_visit_0", session.getNeededDatum().getDataId());
```

Comment:
I think we should add this datum to the frame and confirm that the session also successfully moves on (just to ensure the end-to-end is also still working)